### ISSUE TITLE
Assigning `cls()` does not narrow optional type

### ISSUE THREAD

**Issue by matejcik** [microsoft/pyright#2813]
Using pyright **1.1.203**.
The problem is a regression, wasn't there in 1.1.202.

The following code will **not** typecheck:
```python
class A:
    instance: "A" | None = None

    def __init__(self) -> None:
        self.foo = True

    @classmethod
    def get_foo(cls) -> bool:
        if cls.instance is None:
            cls.instance = cls()
        reveal_type(cls.instance)  # info: Type of "cls.instance" is "A | None"
        return cls.instance.foo  # error: "foo" is not a known member of "None"
```

Changing the assignment to `cls.instance = A()` makes the error go away.

In pyright .202, the revealed type of `cls.instance` is `Self@A | A`.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.204, which I just published. It will also be included in the next release of pylance.
